File: scala/99.scala
object Solution {
    def recoverTree(root: TreeNode): Unit = {
        var x:TreeNode = null
        var y:TreeNode = null
        var prev: TreeNode = null
        def inorder(root: TreeNode) {
          if (root == null) return
          inorder(root.left)
          if (prev != null && root.value < prev.value) {
            x = root
            if (y==null)
              y = prev
            else return
          }
          prev = root
          inorder(root.right)
        }
        inorder(root)
        def swap(x: TreeNode, y: TreeNode) {
          val tmp = x.value
          x.value = y.value
          y.value = tmp
        }
        swap(x, y)
    }
}

File: leetcodeJava/99.java
package leetcodeJava;

class Solution {
    public void recoverTree(TreeNode root) {
        dfs(root);
        var tmp = first.val;
        first.val = second.val;
        second.val = tmp;
    }
    
    private TreeNode prev = null;
    
    private TreeNode first = null, second = null;
    
    private void dfs(TreeNode root) {
        if (root != null) {
            dfs(root.left);
            if (prev == null) {
                prev = root;
            } else {
                if (root.val < prev.val) {
                    if (first == null) {
                        first = prev;
                    }
                    second = root;
                }
                prev = root;
            }
            dfs(root.right);
        }
    }
}
